base_layer/core/src/base_node/sync/header_sync/synchronizer.rs

```diff
@@ -245,59 +249,69 @@ impl<'a, B: BlockchainBackend + 'static> HeaderSynchronizer<'a, B> {
             latency.unwrap_or_default().as_millis()
         );
 
-        // Fetch the local tip header at the beginning of the sync process
-        let local_tip_header = self.db.fetch_last_chain_header().await?;
-        let local_total_accumulated_difficulty = local_tip_header.accumulated_data().total_accumulated_difficulty;
-        let header_tip_height = local_tip_header.height();
-        let sync_status = self
-            .determine_sync_status(sync_peer, local_tip_header, &mut client)
+        // Fetch best local data at the beginning of the sync process
+        let best_block_metadata = self.db.get_chain_metadata().await?;
+        let best_header = self.db.fetch_last_chain_header().await?;
+        let best_block = self
+            .db
+            .fetch_chain_header(best_block_metadata.height_of_longest_chain())
+            .await?;
+        let best_header_height = best_header.height();
+        let best_block_height = best_block.height();
+
+        if best_header_height < best_block_height ||
+            best_block_height < self.local_cached_metadata.height_of_longest_chain()
+        {
+            return Err(BlockHeaderSyncError::ChainStorageError(
+                ChainStorageError::CorruptedDatabase("Inconsistent block and header data".to_string()),
+            ));
+        }
+
+        let peer_response = self
+            .find_chain_split(sync_peer.node_id(), &mut client, NUM_INITIAL_HEADERS_TO_REQUEST as u64)
             .await?;
-        match sync_status {
-            SyncStatus::InSync | SyncStatus::WereAhead => {
-                let metadata = self.db.get_chain_metadata().await?;
-                if metadata.height_of_longest_chain() < header_tip_height {
+        let header_sync_status = self
+            .determine_sync_status(&sync_peer.to_string(), best_header, best_block, peer_response.clone())
+            .await?;
+
+        match header_sync_status.clone() {
+            HeaderSyncStatus::InSyncOrAhead => {
```

Comment:
```suggestion
            HeaderSyncStatus::InSyncOrAhead{ peer_response } => {
```
Rather propagate it here

Reply:
As above